### PR TITLE
Rework external tiles.json

### PR DESCRIPTION
**Pull request by lilleyse**
Continuation of #3237

I tried experimenting with different approaches to loading external `tiles.json` that makes use of `Tileset3DTileContentProvider`. I was able to clean it up a good amount, so now `selectTiles` looks a lot more like the original code, with this addition:
https://github.com/AnalyticalGraphicsInc/cesium/blob/bd97bc4f9fb2733536c3b810084fcff6510a700e/Source/Scene/Cesium3DTileset.js#L400-L409

I'm not handling the case where replacement refinement should select its parent if the subtree isn't  ready, but I don't think it's hard to add.

@pjcozzi and @e-andersson: Can you look at this and make sure I'm not forgetting something?

### CODE REVIEW

File: Source/Scene/Cesium3DTileset.js [hunk 1]
```diff
@@ -442,26 +395,25 @@ define([
             var k;
             var additiveRefinment = (t.refine === Cesium3DTileRefine.ADD);
```
Comment (e-andersson) [inline Source/Scene/Cesium3DTileset.js]: Missing 'e' in variable name, `additiveRefinment` should be `additiveRefinement`

Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: Ah, thanks for catching that.

**Comment by e-andersson** [comment]
Didn't have time today for checking this more carefully and testing, will do first thing next week.

File: Source/Scene/Cesium3DTile.js [hunk 2]
```diff
@@ -146,8 +146,9 @@ define([
             var contentFactory = Cesium3DTileContentProviderFactory[type];
 
             if (type === 'json') {
+                // TODO : should this be in the factory class?
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTile.js]: > TODO : should this be in the factory class?

Yes, it wasn't before because the parameters and workflow were different.

File: Source/Scene/Tileset3DTileContentProvider.js [hunk 3]
```diff
@@ -12,11 +12,41 @@ define([
     /**
      * @private
      */
-    var Tileset3DTileContentProvider = function() {
+    var Tileset3DTileContentProvider = function(tileset, tile, url) {
```
Comment (pjcozzi) [inline Source/Scene/Tileset3DTileContentProvider.js]: Throughout this branch, in a separate PR, can you change the function declarations to, for example,
```javascript
function Tileset3DTileContentProvider(tileset, tile, url) {
```
See #3329

File: Source/Scene/Cesium3DTileset.js (hunk 1 above)
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: That was probably my fault.

File: Source/Scene/Cesium3DTileset.js [hunk 4]
```diff
@@ -272,6 +226,50 @@ define([
         }
     });
 
+    Cesium3DTileset.prototype.loadTilesJson = function(tilesJson, parentTile) {
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Add a comment above to document this as `@private` since it won't be part of the public API.

File: Source/Scene/Cesium3DTileset.js [hunk 5]
```diff
@@ -440,28 +393,29 @@ define([
...
             var k;
-            var additiveRefinment = (t.refine === Cesium3DTileRefine.ADD);
+            var additiveRefinement = (t.refine === Cesium3DTileRefine.ADD);
+
+            if (t.hasTilesetContent && t.isReady()) {
+                // If tile has tileset content, skip it and process its child instead (the tileset root)
+                child = t.children[0];
+                if (child.isContentUnloaded()) {
+                    if (outOfCore) {
+                        requestContent(tiles3D, child);
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Up to you, but it might be a tad cleaner to pass `outOfCore` to `requestContent` and test it there instead of checking it in three separate places like we do now.

In general, we should try to keep this tree traversal code as clean as possible since it is the most important function in 3D Tiles.

File: Source/Scene/Cesium3DTileset.js [hunk 6]
```diff
@@ -440,28 +393,29 @@ define([
             var childrenLength = children.length;
             var child;
             var k;
-            var additiveRefinment = (t.refine === Cesium3DTileRefine.ADD);
+            var additiveRefinement = (t.refine === Cesium3DTileRefine.ADD);
+
+            if (t.hasTilesetContent && t.isReady()) {
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Add a comment above `requestContent` below that says we don't check the child's visibility or SSE before requesting it because, the child is semantically the same tile (with the same SSE and bounding volume).

**Comment by pjcozzi** [comment]
@lilleyse this is a clean approach.

> I'm not handling the case where replacement refinement should select its parent if the subtree isn't ready, but I don't think it's hard to add.

Let's add support for this before merging.

@e-andersson we'll also wait for your comments before merging.

**Comment by e-andersson** [comment]
This looks much cleaner now. Well done, @lilleyse! :+1: 
I have nothing to add except for a small tweak (#3340) to select the parent tile if the child of an external tile is not yet ready.

**Comment by lilleyse** [comment]
This is ready. @e-andersson I cherry picked your commit from #3340, thanks!

**Comment by lilleyse** [comment]
Actually I just considered a possible edge case here: the tile is using replacement refinement and selects all its children. One of the children has tileset content, and it's child (the subtree root) is not ready. As the result, the first tile is selected. But the rest of its children that didn't have tileset content are still selected.

**Comment by lilleyse** [comment]
A solution might be for `Tileset3DTileContentProvider` to request its child (subtree root) immediately and only resolve the ready promise when its child is ready.

**Comment by pjcozzi** [comment]
> A solution might be for Tileset3DTileContentProvider to request its child (subtree root) immediately and only resolve the ready promise when its child is ready.

Sounds good.

**Comment by lilleyse** [comment]
This goes beyond `Tileset3DTile`. The same empty space problem with happen if a tile has a `Empty3DTile` child, which itself has a child with content.

**Comment by e-andersson** [comment]
I stumbled on that problem when trying to think of a way to implement this if we were to allow nested external tiles. Is there any way we could solve this without traversing a whole potential subtree?

**Comment by lilleyse** [comment]
I was considering something like this:

```
Tileset3DTileContentProvider.prototype.request = function() {
    var that = this;

    this.state = Cesium3DTileContentState.LOADING;

    this._tileset.loadTilesJson(this._url, this._tile).then(function() {
        that.state = Cesium3DTileContentState.PROCESSING;
        that.processingPromise.resolve(that);
        
        // Only return ready promise when its child is ready (or loop over all children)
        var child = that.tile.children[0];
        child.requestContent();
        child.readyPromise.then(function() {
            that.state = Cesium3DTileContentState.READY;
            that.readyPromise.resolve(that);
        }).otherwise(function(error) {
            that.state = Cesium3DTileContentState.FAILED;
            that.readyPromise.reject(error);
        });
    }).otherwise(function(error) {
        that.state = Cesium3DTileContentState.FAILED;
        that.readyPromise.reject(error);
    });
};
```

`Empty3DTileContentProvider` could do something similar. The problem here is that it's circumventing the request scheduler.

**Comment by pjcozzi** [comment]
If I am following this correctly, I think it could be solved by keeping a `refinable` flag on each tile (totally open to a better name), which is `true` when it is safe to refine to that tile.

So in the case when there are empty tiles used just for culling before a tile with content, once the tile's content loads, it would walk up the tree setting `refinable` to `true`.

Also, we could probably replace the flag with a new state like splitting `READY` into `READY_TO_TRAVERSE` and `READY_TO_RENDER`, which was discussed in #3237.

**Comment by e-andersson** [comment]
@lilleyse If we used a similar approach in `Empty3DTileContentProvider` I think it wouldn't be possible to use them to make a `tileset of tilesets` structure without having the tilesets in the leaf nodes load.

**Comment by lilleyse** [comment]
Yeah my solution has some problems, @pjcozzi 's seems cleaner.

**Comment by e-andersson** [comment]
It seems we need to know if all children are `refinable` before refining though (at least with replacement refinement). Keeping track of the number of children that are `refinable` vs total number of children, in several layers, may not be easy to do in a clean way. Am I missing something?

**Comment by pjcozzi** [comment]
> It seems we need to know if all children are refinable before refining though (at least with replacement refinement).

Yes, for replacement refinement.  Additive refinement does not.

**Comment by pjcozzi** [comment]
> Keeping track of the number of children that are refinable vs total number of children, in several layers, may not be easy to do in a clean way. Am I missing something?

I dunno.  We probably need more than the flag; we might need something like a `count` that indicates that it is OK to refine once it is zero (or a bitmask with one bit per child).

We can also add limitations to the spec to make the implementation reasonable.

**Comment by lilleyse** [comment]
The code now supports replacement refinement when child tiles are empty. I basically did what @pjcozzi suggested with a `numberOfUnrefinableChildren` parameter (need a better name).

The `selectTiles` traversal isn't too different, mostly extra checks.

**Comment by pjcozzi** [comment]
> `numberOfUnrefinableChildren` parameter (need a better name).

I've tried, but I don't have a better name.

File: Source/Scene/Cesium3DTileset.js [hunk 7]
```diff
@@ -332,51 +336,11 @@ define([
...
-                tile.content.state = Cesium3DTileContentState.READY;
-            });
-        }
-    }
-
     function selectTile(selectedTiles, tile, fullyVisible, frameState) {
         // There may also be a tight box around just the tile's contents, e.g., for a city, we may be
         // zoomed into a neighborhood and can cull the skyscrapers in the root node.
-        if (tile.isReady() &&
+        var parentRefines = defined(tile.parent) ? tile.parent.isRefinable() : true;
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: A better name might be `refinableParent`.

Can you also add a comment here that explains why this is checked, e.g., because we don't want to select a tile for replacement refinement unless all the tiles needed to replace the parent are ready.

Also, do we only need this check for replacement refinement?  Does this always pass for additive refinement?

**Comment by pjcozzi** [comment]
Any recommended spec updates?

**Comment by pjcozzi** [comment]
Any concern about corner cases such as a mix of additive and replacement refinement in the same tree?

File: Source/Scene/Cesium3DTileset.js [hunk 8]
```diff
@@ -528,20 +484,20 @@ define([
...
 
                         if (outOfCore) {
                             for (k = 0; (k < childrenLength) && requestScheduler.hasAvailableRequests(); ++k) {
                                 child = children[k];
 // TODO: we could spin a bit less CPU here and probably above by keeping separate lists for unloaded/ready children.
                                 if (child.isContentUnloaded()) {
-                                    requestContent(tiles3D, child);
+                                    requestContent(tiles3D, child, outOfCore);
+                                } else if (!child.isRefinable()) {
+                                    // Child needs to load its children in order to become refinable
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Can we expand on this comment a bit, e.g., "needs to load its children or, in the case of empty child tiles used to accelerate culling, its ancestors with content need to be loaded"

**Comment by pjcozzi** [comment]
That's all my comments.

In the test branch, we'll need to carefully test all the possibilities, e.g,
* All children have content.
* No children have content, but all grandchildren have content.
* One child has content, the other child is empty, but its children have content.

File: Source/Scene/Cesium3DTileset.js (hunk 7 above)
Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: We need the check for both, I'll add a comment here to make that clearer.

**Comment by lilleyse** [comment]
I agree this needs more testing, I can already see some problems with it that I need to fix.

**Comment by pjcozzi** [comment]
> I agree this needs more testing, I can already see some problems with it that I need to fix.

I am also fine with merging this into the 3d-tiles branch as long as it doesn't break anything (e.g., if just doesn't fix everything) to aid in your testing.

**Comment by lilleyse** [comment]
I fixed up the algorithm and moved the tree traversal out of `selectTiles`. If this looks okay, I'm fine with merging this into the 3d-tiles branch so I can start writing some tests.

**Comment by pjcozzi** [comment]
Looks good.

When you write the tests for some of the corner cases, consider stepping through the tree traversal in the debugger just to cross all the t's and dot all the i's.

File: Source/Scene/Cesium3DTileset.js [hunk 9]
```diff
@@ -440,28 +397,31 @@ define([
...
+            if (t.hasTilesetContent) {
+                // If tile has tileset content, skip it and process its child instead (the tileset root)
+                // No need to check visibility or sse of the child because its bounding volume
+                // and geometric error are equal to its parent.
+                if (t.isReady()) {
+                    child = t.children[0];
+                    if (child.isContentUnloaded()) {
+                        requestContent(tiles3D, child, outOfCore);
+                    } else {
+                        stack.push(child);
```
Comment (e-andersson) [inline Source/Scene/Cesium3DTileset.js]: @lilleyse The child tile pushed on the stack will have `distanceToCamera === 0` here, so when it comes off the stack `sse` will be calculated based on that. Don't we need to compute tile distance here?

Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: Both good points, I'll make sure to add those changes to #3367.

Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: I made the changes in this commit: 38e29e1
